### PR TITLE
Kubecon 2018

### PR DESCRIPTION
**Pull request by vgallissot**
add 3 articles: 1 for each day.
+ a little review at the end.

Made with love with our hands :heart:

### CODE REVIEW

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+## CNCF 20-20 Vision - Alexis Richardson, Founder & CEO, Weaveworks
+
+
+![Waiting for the Keynotes to start](/images/posts/kubecon-2018/kubecon-before-keynotes.jpg)
+
+
+# Whats Up With All The Different Container Runtimes? - Ricardo Aravena, Branch Metrics
+
+Overview of the different containers runtimes, starting with OpenVZ in 2006 (still maintained, but the last 2.7 version is not as stable as the previous one, and it doesn't support Kubernetes), and LXC (Kubernetes support is low-priority WIP and LXC uses a specific image format) in 2011. Docker (initially based on LXC) arrived in 2013 and things have gone insane since then. With libcontainer and rkt in 2014, OCI in 2015 and CRI in 2016. Today, Kubernetes supports several runtines.
```
Comment (b-viguier) [inline _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: > Kubernetes supports several **runtimes**.

❓

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+rkt could be interesting from a security point of view (supports TPM and VMs). With Kubernetes 1.10, the default runtime is runc. crun is the most interesting runtime for performances, but is WIP and isn't currently used much. Kata, released in 2017 has the best security (it runs containers in VMs) and is OCI compliant, but is slower and more heavyweight. Other very specific runtimes include nvidia, railcar, pouch, lmctfg, systemd-nspawn...
```
Comment (b-viguier) [inline _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: Strange for me to start en sentence without upper case. I guess rkt and crun are some acronyms, so why not style them `rkt` and `crun`? Or using an upper case like Kata? (what about nvidia, railcar… ?)
TL;DR: consistency for the win 😄

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+
+# Seamless Development Environments on Kubernetes using Telepresence - Ara Pulido, Bitnami
+
+Kubernetes is a great production environment, but it feels like development environment is kind of an afterthought: even for a simple application, if you want to develop (locally?), things are not easy. People are currently using two distinct ways: using docker compose to replicate the production environment (but compose doesn't do everything: rbac, job, ingress... and having to maintain everything twice is not fun), or build/push/deploy-to-k8s and wait many seconds everytime one wants to F5 on a page, which is unberably slow (I wouldn't ask my developers' colleagues to do this for even half a day!).
+
+The solution proposed during this conference is [Telepresence](https://www.telepresence.io/). It allows a developer to swap out a pod from a cluster and *inject* her own pod, running locally, at its place. Some sort of VPN is established between her computer and the cluster, which means the pod running locally behaves just like if it was still in the cluster (including DNS, service discovery, access to non-Kubernetes managed services and all).
```
Comment (b-viguier) [inline _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: > It allows developer to … **her** own pod

her? his? Or to be more neutral:
> It allows developer**s** to… **their** own pod

File: _posts/2018-05-22-kubecon-2018-copenhaguen-day-2.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+Autoscaling on Kubernetes used to rely on Heapster. But Heapster's API is unspecified, it doesn't work with Prometheus, and vendor implementations are often unmaintained. Starting with 1.8, Kubernetes provides a resource and custom metrics API. It is a specification (not an implementation), the implementations are developed and of maintained by vendors, and each metric returns a single value. Kubernetes' HPA (HorizontalPodAutoscaler) uses these metrics to scale up/down.
+
+Core metrics are CPU and RAM, by pod/container/node, and there is a canonical implementation called [`metrics-server`](https://github.com/kubernetes-incubator/metrics-server). The custom metrics API has the same semantics (a single value is returned) but no canonical implementation is provided (take a look at [`DirectXMan12/k8s-prometheus-adapter`](DirectXMan12/k8s-prometheus-adapter) for an implementation). Each metric is related to a Kuberntes object (Pod, Service, Deployment, ...). Finally, there is also an external metrics API (currently in alpha stage) for things not related to a Kubernetes object (like the queue length for a queuing service provided by a Cloud provider -- SQS on AWS, for example).
```
Comment (b-viguier) [inline _posts/2018-05-22-kubecon-2018-copenhaguen-day-2.md]: > Each metric is related to a Kubern**e**tes object

File: _posts/2018-05-23-kubecon-2018-copenhaguen-day-3.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+# Operating a Global-Scale FaaS on Top of Kubernetes - Chad Arimura & Matt Stephenson, Oracle
+
+This was about the [Fn project](http://fnproject.io/). A couples of problems related to multitenancy: network isolation on Kubernetes, noisy neighbors (I/O being the bottleneck). Helm has a few limits, worked around with shell scripts.
+
+
+# Inside Kubernetes Resource Management (QoS) – Mechanics and Lessons from the Field - Michael Gasch, VMware
+
+Resource management goes through cgroups. With containers, we see all CPU/RAM, but this doesn't mean we'll be able to use them all: we may have to share with other containers. Work with requests. For now, cpu and memory are stable resources, but others (hugepages, ephemeral storage, device plugins) are in beta. You should align Kubernetes' QoS with the underlying infrastructure, enable quotas in the cluster, and protect critical system pods.
+
+I have to admit I didn't take much notes during this talk, but noted the slides contain a lot of informations -- for more, go read them ;-)
```
Comment (b-viguier) [inline _posts/2018-05-23-kubecon-2018-copenhaguen-day-3.md]: A link to the slides?

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+language: en
+---
+
+At the very beginning of May, we (Pascal and Vincent) went to KubeCon 2018. It was a three-days long event, with more than 300 conferences. 4300 people met at Bella Center, a huge conference place close to Copenhaguen in Denmark. Here are some of our notes about some talks we saw.
+
+
+# Keynotes
+
+## How Good Is Our Code? - Dan Kohn, Executive Director, Cloud Native Computing Foundation
+
```
Comment (kennydee) [inline _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: This is WIP ? 
If not, it is better to remove those without any comment i think

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+## CNCF Project Update - Liz Rice, Technology Evangelist, Aqua Security; Sugu Sougoumarane, CTO, PlanetScale Data; Colin Sullivan, Product Manager, Synadia Communications, Inc. & Andrew Jessup, Co-founder, Scytale Inc.
+
+During this keynote, I realized the CNCF is helping a lot more projects than I thought it was: it's not just Kubernetes. I will take a closer look to some of them in the future -- they are all listed on [l.cncf.io](https://landscape.cncf.io/).
```
Comment (kennydee) [inline _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: CNCF ?

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md (hunk 2 above)
Comment (kennydee) [reply _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: yes, use backquote for "software name" please

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+rkt could be interesting from a security point of view (supports TPM and VMs). With Kubernetes 1.10, the default runtime is runc. crun is the most interesting runtime for performances, but is WIP and isn't currently used much. Kata, released in 2017 has the best security (it runs containers in VMs) and is OCI compliant, but is slower and more heavyweight. Other very specific runtimes include nvidia, railcar, pouch, lmctfg, systemd-nspawn...
+
+Basically, today and for most workloads, you should go with the standard docker/containerd runtime. There is a convergence towards OCI, it's the default for Kubernetes and docker is going to adopt the CRI plugin.
+
+
+# Introduction to Istio Configuration - Joy Zhang, Google
+
+An introduction to the [Istio Service Mesh](https://istio.io/). All Istio components are CRDs. CRDs are becoming a standard when it comes to kubernetes customizations, here requests proxying.
```
Comment (kennydee) [inline _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: CRD ? i like, when acronyms are defined the first time we talk about them

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+
+# How to Export Prometheus Metrics from Just About Anything - Matt Layher, DigitalOcean
+
+This presentation was about a few good practices to follow when it comes to exposing Prometheus metrics from a Go application. Basically, you should use the Go client library, be really careful about concurrency, build reusable packages, write unit-tests, use `promtool check-metrics`, and read and follow the Prometheus metrics best practices.
+
+
+# Continuously Deliver your Kubernetes Infrastructure - Mikkel Larsen, Zalando SE
+
+Another really good rex from Zalando from their utilization of k8s in prod and lessons learned.
```
Comment (kennydee) [inline _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: did REX means something in english ?

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+*        That means Jenkins will be able to run CI tests, Continuously Deploy your project to preview, staging, prod and so on with Skaffold/Helm to k8s
+*        Jenkins will run pipelines from the `Jenkinsfile` in the repo to do that CI/CD part
+*        In the provided pipelines given with `jx import`, you will use provided docker images that embed `jx` cli and other tools to manage the deployments of your app.
+*        That allows you to promote your app between stages, build your docker image, etc. in your pipeline steps.
+*        Those deployments are based on Helm Charts in the repo.
+*        Jenkins-x follows *GitOps* strategy, that means anything useful is stored in each app's repo: it is versioned and git events will trigger pipelines.
+
+Jenkins-X brings this CI/CD part that was missing for k8s users. Gitlab + gitlab-ci were already doing that for some years now, but nothing was that fancy for GH users.
+
+We are very excited about Jenkins-X, as it answers a lot of problematics and brings in gitops as core concept. We're actively adopting it and we hope to give feedback asap.
```
Comment (kennydee) [inline _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: Spoiler alert ! Love that :)

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+
+
+# Keynote: Anatomy of a Production Kubernetes Outage - Oliver Beattie, Head of Engineering, Monzo Bank
+
+This keynote was about  major outage at Monzo. You can read more about it in the [post-mortem](https://community.monzo.com/t/resolved-current-account-payments-may-fail-major-outage-27-10-2017/26296/94) they posted after it happened. Basically, even when you are careful, an outage can still happen: several causes combined with a very specific bug happening with specific versions in a specific case and voilà. Nice talk, and nice to hear a bank being so open!
```
Comment (kennydee) [inline _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: two spaces after `about` at the beggining of the sentence

File: _posts/2018-05-22-kubecon-2018-copenhaguen-day-2.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+## Switching Horses Midstream: The Challenges of Migrating 150+ Microservices to Kubernetes - Sarah Wells, Technical Director for Operations and Reliability, Financial Times
+
+## Shaping the Cloud Native Future - Abby Kearns, Executive Director, Cloud Foundry Foundation
+
+## Skip the Anxiety Attack - Build Secure Apps with Kubernetes - Jason McGee, Fellow, IBM
+
+## Software's Community - Dave Zolotusky, Software Engineer, Spotify
+
+
+# Autoscale your Kubernetes Workload with Prometheus - Frederic Branczyk, CoreOS
```
Comment (kennydee) [inline _posts/2018-05-22-kubecon-2018-copenhaguen-day-2.md]: > SLO of SLA through SLI 
:D you'll need to explain that a little bit more i think

File: _posts/2018-05-22-kubecon-2018-copenhaguen-day-2.md [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+
+# Clusters as Cattle: How to Seamlessly Migrate Apps across Kubernetes Clusters - Andy Goldstein, Heptio
+
+As many people, Andy has a lot of clusters. To re-route traffic between clusters, he uses envoy.
+To maintain consistent configurations, he uses ansible to provision everything.
+So far, I don't really see the point of having a lot of clusters and even less of migrating a single app between clusters, but that can be interesting for people that like that trend.
+
+
+# Horizontal Pod Autoscaler Reloaded - Scale on Custom Metrics - Maciej Pytel, Google & Solly Ross, Red Hat
```
Comment (kennydee) [inline _posts/2018-05-22-kubecon-2018-copenhaguen-day-2.md]: You could delete this talk.
I think `less is better`, so you can remove those where you don't have anything to say about.

File: _posts/2018-05-23-kubecon-2018-copenhaguen-day-3.md [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+# Integrating Prometheus and InfluxDB - Paul Dix, InfluxData
+
+I was a lot surprised in this talk as we were like 30 in a room for 300.
+I guess people don't need to keep metrics more than 15days with the Prometheus engine, or maybe people are already using InfluxDB as a guaranteed long term storage.
+Because we are in the second case, we are testing influxDB with Prometheus, so this talk came at the right time.
+I didn't learn much on InfluxDB + Prometheus, nor on federated queries that comes with HA.
+
+Paul then questioned why not use a single query language for all query engines? That would be more
+practical and maintainable. The question is still open, even if Paul proposes IFQL to rule them all.
```
Comment (kennydee) [inline _posts/2018-05-23-kubecon-2018-copenhaguen-day-3.md]: unwanted carriage return ?

File: _posts/2018-05-23-kubecon-2018-copenhaguen-global-review.md [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+First of all "Cloud native" seems to be *the* buzzword of the year. Not just *cloud* anymore, but *cloud native*! What does it mean? Instead of just deploying your application to the cloud, it should fully *use the cloud*.
+
+Then, Kubernetes. It is a mature solution in itself. There doesn't seem to be any doubt left about that. Even if this is the case with Kubernetes, the majority of the ecosystem around it is not mature yet, and that's a bit of a problem. We saw a lot of tools and some that were presented during talks are still WIP and some demos completely failed. If you're surfing the Kubernetes wave, please be careful with the tools you choose, and don't loose yourself adopting a fancy/non-working tool that will bring your infrastructure down. Continue to master what you do without being trapped by the hype brought by certain solutions.
+
+Deployment, CI and CD. Well, not so much. There are a few projects out there and several different approaches (`kubectl apply`, a bit of Jenkins around it, deployments from inside the cluster, several black boxes like CodeFresh ...), but not one thing that everyone is doing/using.
```
Comment (kennydee) [inline _posts/2018-05-23-kubecon-2018-copenhaguen-global-review.md]: Talk a little bit about jenkinX here too ?

**Comment by kennydee** [comment]
And oh i forgot, if you could do at the end of the process review, three PR (one by article) to merge one each day for exemple, it could be great :) (if sysadmin can do cherry-pick :trollface: ?)

File: _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md (hunk 3 above)
Comment (vgallissot) [reply _posts/2018-05-21-kubecon-2018-copenhaguen-day-1.md]: We had the discussion over this with Pascal already :D
Not not always use **her**, **his** or **their**, we mix them all.
The idea is the same, but that enhance the reader's experience.
